[WebGPU] Range check built-in functions

tadeuzagallo opened the pull request:
#### bb9256b06c27e21ce833aa01a41d103cb0e97b59
<pre>
[WebGPU] Range check built-in functions
<a href="https://bugs.webkit.org/show_bug.cgi?id=289776">https://bugs.webkit.org/show_bug.cgi?id=289776</a>
<a href="https://rdar.apple.com/147019863">rdar://147019863</a>

Reviewed by Mike Wyrzykowski.

Introduce __wgsl builtins for all trigonometric functions for bounds checking the input.
For every function `f` the builting does `f(select(value, 0, value &lt; -1 || value &gt; 1))`.

The patch also introduces a new mechanism to generate helpers, that doesn&apos;t require
manually tracking what helpers are used in other compilation phases. Later all simple
helpers should be refactored to use this mechanism.

* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
(WGSL::Metal::FunctionDefinitionWriter::FunctionDefinitionWriter):
(WGSL::Metal::FunctionDefinitionWriter::stringBuilder):
(WGSL::Metal::FunctionDefinitionWriter::write):
(WGSL::Metal::FunctionDefinitionWriter::emitNecessaryHelpers):
(WGSL::Metal::FunctionDefinitionWriter::visit):
(WGSL::Metal::FunctionDefinitionWriter::generatePackingHelpers):
(WGSL::Metal::FunctionDefinitionWriter::emitPackedVector):
(WGSL::Metal::FunctionDefinitionWriter::serializeVariable):
(WGSL::Metal::FunctionDefinitionWriter::visitArgumentBufferParameter):
(WGSL::Metal::FunctionDefinitionWriter::serializeBinaryExpression):
(WGSL::Metal::FunctionDefinitionWriter::visitStatements):
(WGSL::Metal::FunctionDefinitionWriter::serializeConstant):
* Source/WebGPU/WGSL/wgslc.cpp:
(CommandLine::parseArguments):
(runWGSL):

Canonical link: <a href="https://commits.webkit.org/292296@main">https://commits.webkit.org/292296@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3e5c9f4bd61f06a207cb29fe8b26f7231f810dfb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/95496 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/15098 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/4981 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/100544 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/46001 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/15383 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/23529 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/72834 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30103 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/98499 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/11524 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/86241 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53166 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/11230 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/3952 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/45337 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/81420 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/4071 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/102580 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/22546 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/16463 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/81874 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/22798 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/82252 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81225 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20356 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/25805 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/3265 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/15877 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/22514 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/27670 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/22173 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/25649 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/23915 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -48,10 +48,50 @@ namespace Metal {
+#define DEFINE_TRIG_HELPER(__name, __capitalizedName, ...) \
+    DEFINE_HELPER(__capitalizedName,  \
+    template <typename T> \
+    T __wgslAsin(T value) \
```
does `__wgslAsin` need to be `__wgsl##__capitalizedName` or am I misreading the macro?

mwyrzykowski replied:
indeed this is observed in https://webgpu.github.io/webgpu-samples/?sample=particles

tadeuzagallo replied:
oops, such a silly mistake, thanks for catching it.

mwyrzykowski commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -48,10 +48,50 @@ namespace Metal {
+    template <typename T> \
+    T __wgslAsin(T value) \
+    { \
+        return __name(select(value, T(0), value < T(-1) || value > T(1))); \
```
I think the condition:
```
value < T(-1) || value > T(1))
```
should be replaced with:
```
value < T(__lowerBound) || value > T(__upperBound))
```

where `DEFINE_TRIG_HELPER(__name, __capitalizedName, __lowerBound, __upperBound, ...) \`

and then:
```
DEFINE_TRIG_HELPER(acos, Acos, -1, 1)
DEFINE_TRIG_HELPER(asin, Asin, -1, 1)
DEFINE_TRIG_HELPER(atan, Atan, numeric_limits<T>::lowest(), numeric_limits<T>::max())
DEFINE_TRIG_HELPER(cos, Cos, numeric_limits<T>::lowest(), numeric_limits<T>::max())
DEFINE_TRIG_HELPER(sin, Sin, numeric_limits<T>::lowest(), numeric_limits<T>::max())
DEFINE_TRIG_HELPER(tan, Tan, numeric_limits<T>::lowest(), numeric_limits<T>::max())
DEFINE_TRIG_HELPER(acosh, Acosh, 1, numeric_limits<T>::max())
DEFINE_TRIG_HELPER(asinh, Asinh, numeric_limits<T>::lowest(), numeric_limits<T>::max())
DEFINE_TRIG_HELPER(atanh, Atanh, -1, 1)
DEFINE_TRIG_HELPER(cosh, Cosh, numeric_limits<T>::lowest(), numeric_limits<T>::max())
DEFINE_TRIG_HELPER(sinh, Sinh, numeric_limits<T>::lowest(), numeric_limits<T>::max())
DEFINE_TRIG_HELPER(tanh, Tanh, numeric_limits<T>::lowest(), numeric_limits<T>::max())
```

(updated comment with table)
I suppose all the functions with lower + upper bounds of `numeric_limits<T>::lowest(), numeric_limits<T>::max())` could be omitted from here

mwyrzykowski replied:
The domain is defined https://www.w3.org/TR/WGSL/#acos-builtin

mwyrzykowski reviewed (commented):
In addition to the trig functions, should we define wrappers for all the other WGSL functions which require a specific domain or is that for a separate PR?

E.g., 
https://www.w3.org/TR/WGSL/#inverseSqrt-builtin
`inverseSqrt(x)`; x ∈ [0, +inf]
`log(x)`; x ∈ [0, +inf]
`log2(x)`; x ∈ [0, +inf]
``pow(base, exp)`; limits `base` and `exp` specified: https://en.cppreference.com/w/cpp/numeric/math/pow
> If base is finite and negative and exp is finite and non-integer, a domain error occurs and a range error may occur.
If base is zero and exp is zero, a domain error may occur.
If base is zero and exp is negative, a domain error or a pole error may occur.

`sqrt(x)`; x ∈ [0, +inf]

mwyrzykowski reviewed (commented):
https://webgpu.github.io/webgpu-samples/?sample=normalMap is incorrect after this change, I'm trying to identify the reason why

mwyrzykowski commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -48,10 +48,49 @@ namespace Metal {
+DEFINE_BOUND_HELPER(asin, Asin, -1, 1)
+DEFINE_BOUND_HELPER(acosh, Acosh, 1, numeric_limits<T>::max())
+DEFINE_BOUND_HELPER(atanh, Atanh, -1, 1)
+DEFINE_BOUND_HELPER_RENAMED(inverseSqrt, InverseSqrt, sqrt, 0, numeric_limits<T>::infinity())
```
I think `sqrt` should be `rsqrt` on the above line

mwyrzykowski commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -2038,54 +2081,75 @@ void FunctionDefinitionWriter::visit(const Type* type, AST::CallExpression& call
+            { "reverseBits"_s, NOOP_HELPER(reverse_bits) },
+            { "round"_s, NOOP_HELPER(rint) },
+            { "sign"_s, NOOP_HELPER(__wgslSign) },
+            { "sqrt"_s, NOOP_HELPER(Sqrt) },
```
I think the above line should be either `NOOP_HELPER(__wgslSqrt)` or `EMIT_HELPER(Sqrt)` otherwise the generated MSL has calls to `Sqrt` which does not exist

mwyrzykowski replied:
Noticed running https://webgpu.github.io/webgpu-samples/?sample=cornell

tadeuzagallo replied:
Thanks for catching, will re-upload

mwyrzykowski reviewed (approved):
Thank you!

tadeuzagallo committed:
[WebGPU] Range check built-in functions
https://bugs.webkit.org/show_bug.cgi?id=289776
rdar://147019863

Reviewed by Mike Wyrzykowski.

Introduce __wgsl builtins for all trigonometric functions for bounds checking the input.
For every function `f` the builting does `f(select(value, 0, value < -1 || value > 1))`.

The patch also introduces a new mechanism to generate helpers, that doesn't require
manually tracking what helpers are used in other compilation phases. Later all simple
helpers should be refactored to use this mechanism.

* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
(WGSL::Metal::FunctionDefinitionWriter::FunctionDefinitionWriter):
(WGSL::Metal::FunctionDefinitionWriter::stringBuilder):
(WGSL::Metal::FunctionDefinitionWriter::write):
(WGSL::Metal::FunctionDefinitionWriter::emitNecessaryHelpers):
(WGSL::Metal::FunctionDefinitionWriter::visit):
(WGSL::Metal::FunctionDefinitionWriter::generatePackingHelpers):
(WGSL::Metal::FunctionDefinitionWriter::emitPackedVector):
(WGSL::Metal::FunctionDefinitionWriter::serializeVariable):
(WGSL::Metal::FunctionDefinitionWriter::visitArgumentBufferParameter):
(WGSL::Metal::FunctionDefinitionWriter::serializeBinaryExpression):
(WGSL::Metal::FunctionDefinitionWriter::visitStatements):
(WGSL::Metal::FunctionDefinitionWriter::serializeConstant):
* Source/WebGPU/WGSL/wgslc.cpp:
(CommandLine::parseArguments):
(runWGSL):

Canonical link: https://commits.webkit.org/292296@main
